### PR TITLE
Add description of standard hooks

### PR DESCRIPTION
**Pull request by elezar**
This change adds a description of standard hooks to the CDI specification. These hooks can be used to update the LDCache in the container or to create symlinks.

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -245,6 +245,53 @@ The `containerEdits` field has the following definition:
...
+
+### create-symlinks
+
+The `create-symlinks` hook is a `createContainer` hooks that is used to ensure that
+required symlinks exist in a container. Typically these symlinks point to injected
+libraries or executables.
+
+Assuming the following hook syntax:
+```shell
+[command-prefix] create-symlinks [--link target::link-path]
```
Comment (elezar) [inline SPEC.md:287]: I'm not 100% happy with the `target::link-path` format here. Are there other suggestions on how we would do this?

Something like `--link source=foo,target=bar` instead?

Comment (klihub) [reply SPEC.md:287]: Cannot we use an ln:ish syntax of `--link $existing_target $new_link_name` ?

Comment (reidpr) [reply SPEC.md:287]: I agree with @elezar's initial concern. The way I'd frame it is, because the hook arguments is a JSON list, more syntax within those list elements introduces a new DSL.

The `foo::bar` notation is new to me. I'm guessing it's more standard in golang world, but in Python or C it would require a second stage of argument parsing.

My suggestions would be:

* Adopt a general principle of one value per argument (i.e., no `::` operator).
* Specifically, use @klihub's suggestion of separate target and source arguments.
* Remove `--link` because it's redundant with the hook definition. This would also make argument parsing in Python or C easier because the target and source become positional arguments rather than options for `--link`. (Also yes, an argument that takes multiple options is hard to parse in my world too.) `nvidia-ctk-hook` could simply ignore `--link` to be backwards compatible.
* Make at least one target and one source required.
* Clarify whether the hook takes multiple target/source pairs or just one.

**Review by reidpr — COMMENTED** [review]
Thanks for requesting my opinion.

I think this is a big step forward for the standard.

It does still seem roundabout to me. That is, with the introduction of standard hooks, we now have prescribed behavior, which is great, but it's indirected through the constraints and syntax of command-line programs. To me it would be simpler to just introduce new `updateLDCache` and `createSymlinks` fields.

Further suggestions in-line. I hope this is helpful.

File: SPEC.md [hunk 2]
```diff
@@ -245,6 +245,53 @@ The `containerEdits` field has the following definition:
...
 
+## Hooks
+
+Hooks allow CDI spec vendors to inject specific logic into various points of the
+container lifecycle. These are typically mapped to [OCI runtime hooks](https://github.com/opencontainers/runtime-spec/blob/main/config.md#posix-platform-hooks) and
+are typically used when behaviour depends on the container contents in some way.
+
+The following named hooks are defined by the CDI specification:
+* `update-ldcache`
+* `create-symlinks`
```
Comment (klihub) [inline SPEC.md:256]: @elezar One question related to this... Is the idea with this document addendum, to describing de-facto hooks, including their command line options and expected behavior ? Also, is the `[command prefix] $named_hook` syntax to allow for an implementation, where a single binary implements all the de-facto hooks (and an extra 1st $hook_name argument is used to pick which hook to run from that binary) ?

If it is so, and the CDI specification provides such a detailed description of these de-facto hooks then wouldn't it make also sense to provide a default implementation for these hooks ?

File: SPEC.md (hunk 1 above)
Comment (elezar) [reply SPEC.md:287]: One thing that's not in the syntax above is that `--link` can be repeated meaning that one can create multiple links with a single invocation of the `create-symlink` command.

Comment (bart0sh) [reply SPEC.md:287]: @elezar How would it look in the CDI file and how it would differ from the currently available approach:
```
"hooks": [
    {"startContainer": {"path": "/usr/bin/ln", "args": ["/src", "/tgt"] }
]

```

File: SPEC.md [hunk 3]
```diff
@@ -250,6 +250,106 @@ The `containerEdits` field has the following definition:
...
+### create-symlinks
+
+The `create-symlinks` hook is a `createContainer` hook that is used to ensure that
+required symlinks exist in a container. Typically these symlinks point to injected
+libraries or executables.
+
+The hooks will have the following YAML definition:
+```yaml
+  hooks:
+  - path: <comnand-prefix>
```
Comment (klihub) [inline SPEC.md:279]: typo: comnand-prefix should be command-prefix

File: SPEC.md [hunk 4]
```diff
@@ -250,6 +250,106 @@ The `containerEdits` field has the following definition:
...
+
+The `update-ldcache` hook is a `createContainer` hook that is used to ensure that
+the ldcache in a container is updated to include any injected libraries. This
+ensures that a user does not have to update environment variables such as `LD_LIBRARY_PATH`
+themselves. The hook also creates `SONAME` symlinks for any injected libraries.
+
+The hooks will have the following YAML definition:
+```yaml
+  hooks:
+  - path: <comnand-prefix>
```
Comment (klihub) [inline SPEC.md:321]: typo: ditto here

**Review by klihub — CHANGES_REQUESTED** [review]
I have two typo-nits that should be fixed.
